pkl-core/src/main/java/org/pkl/core/runtime/VmUtils.java

```diff
@@ -298,6 +298,9 @@ public static Object doReadMember(
           }
         }
       } else if (receiver instanceof VmListingOrMapping<?> vmListingOrMapping) {
+        if (owner != receiver && owner instanceof VmListingOrMapping<?> vmListingOrMappingOwner) {
+          ret = vmListingOrMappingOwner.typecastObjectMember(member, ret, callNode);
+        }
         ret = vmListingOrMapping.typecastObjectMember(member, ret, callNode);
       }
```

Comment:
If the following suggestion doesn't make sense, a test is missing. :-)

```suggestion
      } else if (owner instanceof VmListingOrMapping<?> vmListingOrMappingOwner) {
        ret = vmListingOrMappingOwner.typecastObjectMember(member, ret, callNode);
      }
```